packages/core/assistant/src/session/session.ts

```diff
@@ -28,83 +27,36 @@ import { isNotFalsy } from '@dxos/util';
 import { type AiAssistantError } from '../errors';
 
 import { formatSystemPrompt, formatUserPrompt } from './format';
+import { GenerationObserver } from './observer';
 
-export type AiSessionOptions = {};
-
-/**
- * Live observer of the generation process.
- */
-export interface GenerationObserver {
-  /**
-   * Unparsed content block parts from the model.
-   */
-  onPart: (part: AiResponse.Part) => Effect.Effect<void>;
-
-  /**
-   * Parsed content blocks from the model.
-   * NOTE: Use block.pending to determine if the block was completed.
-   * For each block this will be called 0..n times with a pending block and then once with the final state of the block.
-   *
-   * Example:
-   *  1. { pending: true, text: "Hello"}
-   *  2. { pending: true, text: "Hello, I am a"}
-   *  3. { pending: false, text: "Hello, I am a helpful assistant!"}
-   */
-  onBlock: (block: ContentBlock.Any) => Effect.Effect<void>;
-
-  /**
-   * Complete messages fired during the session, both from the model and from the user.
-   * This message will contain all message blocks emitted through the `onBlock` callback.
-   */
-  onMessage: (message: DataType.Message) => Effect.Effect<void>;
-}
+export type AiSessionRunError = AiError.AiError | AiInputPreprocessingError | AiToolNotFoundError | AiAssistantError;
 
-export const GenerationObserver = Object.freeze({
-  make: ({
-    onPart = Function.constant(Effect.void),
-    onBlock = Function.constant(Effect.void),
-    onMessage = Function.constant(Effect.void),
-  }: Partial<GenerationObserver> = {}): GenerationObserver => ({
-    onPart,
-    onBlock,
-    onMessage,
-  }),
+export type AiSessionRunRequirements<Tools extends AiTool.Any> =
+  | AiTool.ToHandler<Tools>
+  | AiLanguageModel.AiLanguageModel
+  | ToolResolverService
+  | ToolExecutionService
+  | TracingService;
 
-  noop: () => GenerationObserver.make(),
+export type AiSessionRunEffect<Tools extends AiTool.Any> = Effect.Effect<
```

Comment:
I agree with extracting Error and Requirements but I think typedeffing effect is generally an anti-pattern

Reply:
Its the same as typing `type SessionRunPromise = Promise<DataType.Message[]>` vs `Promise<SessionRunResult>`